Pull request for #20: Implement preliminary template comments

yuchuan2016 opened the pull request:
Hi @lboxell ,

Could you PR this? The main changes are in [this comment](https://github.com/gslab-econ/template/issues/20#issuecomment-260723272). They are all minor changes. One thing to note is that I also change the example (in `source/data/` and `source/analysis/`) just to fabricate an example for `source/table/`.

Thanks a lot!

lboxell commented:
@yuchuan2016, I'm having trouble running the  directory with scons. I get 
```
siepr-mg-ra-imac-1:template leviboxell$ scons
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
scons: *** [release/data.txt] Source `build/data/data.txt' not found, needed by target `release/data.txt'.
scons: building terminated because of errors.
``` 

on a new build.

yuchuan2016 commented:
@lboxell , I'm sorry. The error has been corrected. Thanks!

lboxell commented on README.md:
```diff
@@ -1,42 +1,34 @@
+ - [gslab_tools](https://pypi.python.org/pypi/GSLab_Tools) version 1.1.1 or later
+    - Installed via [pip](https://pip.pypa.io/en/stable/) using `pip install gslab_tools`.
+ - [gslab_stata tools](https://github.com/gslab-econ/gslab_stata)
+    - Installed according to [computer-build-sheet](https://github.com/gslab-econ/admin/blob/cd4ddcf91cbd724a0ac444adf18bdba797457e3c/computer-build-sheet/package_build.sh).
```
Is there a way to easily remove reliance on gslab_stata so that the template only relies on public resources? Or we should think about making gslab_stata public.

It will probably be good to propose a minimum functioning template for the Stata script that we want to use going forward and have MG/JMS approve it. Things to consider are whether we want to continue using `preliminaries` and other GSLAB tools by default.

yuchuan2016 replied:
In the current do file, I'm using `matrix_to_txt` command to create an table output used by `Tablefill`, and create `output/tables/table.lyx`. If we want to remove GSLab Stata tools, it will take a few more lines to create a table.txt in the right format.

lboxell replied:
@gentzkow @jmshapir, 

Currently, we use gslab_stata, which is a private repo, in our template. My understanding is that we want the template in a form that only requires public tools. This means we can either remove any reliance on gslab_stata in the template or make gslab_stata public. Any preference?

jmshapir replied:
@lboxell how important is the dependence on gslab_stata? If it's not that important it might be cleaner to remove it.

lboxell replied:
@jmshapir,

My understanding is that we currently just use gslab_stata to create a table for tablefill. We could either add a few more lines of code to the stata code in the template to produce a proper table without using matrix_to_txt in gslab_stata or we could use a different program to create the table.

jmshapir replied:
@lboxell if it is easy to do I would support removing the dependency on matrix_to_txt since that's not needed to illustrate the functionality of the template.

lboxell replied:
Sounds good.

lboxell commented on README.md:
```diff
@@ -1,42 +1,34 @@
  - [Lyx](https://www.lyx.org/)
-    - Executable should also be added to the `PATH` environment variable so it is callable from the command line with `lyx`
+    - For the above four, executables need to added to `path` environment variable. So they are callable from the command line with `python`, `statamp`, `Rscript`, and `lyx`.
+ - [SCons](http://scons.org/) version 2.4.0 or later
```
Change to (since one can use version 2.4.0 or below and still run the repo):

- [SCons](http://scons.org/)
  - version 2.4.0 or later is best if using the cache.

yuchuan2016 commented:
Hi @lboxell ,

Thanks a lot! As to the setup.py, I'm not quite sure if I'm on the right way. I have written [this gist](https://gist.github.com/yuchuan2016/1b113b5528f729f2b0a70cd157f02dde). Do you think we want something like this? Thanks!

lboxell commented:
Per discussion, I think the gist is the right direction. For this task, we can just focus on making things Mac compatible and then have a follow up task at some point to make things windows compatible (and possibly move setup.py to gslab_python if it starts getting too large).

yuchuan2016 commented:
@lboxell ,

I have removed the `output` folder and added a `setup.py` at the root of the repo. But I'm not quite sure whether the `release` protocol is correct. I only release the paper and slide, and other outputs are stored in `build` not versioned.

lboxell commented:
@yuchuan2016,

I made some edits (mostly to the setup.py and the readme). Let me know if you agree with them. 

If so, I think we should to the following:
 - Merge `master` into this branch and make any needed edits.
 - Remove the `check_os()` call and the mention of requiring Mac/Linux OS in the readme and ensure things still run on Windows. (The only thing that I see an issue with is the `check_metropolis()` command will need to have a windows option.)
 - Merge into `master`
 - Create a task in gslab_python to remove any redundancies between `setup.py` and `gslab_scons`. (I think this will mostly relate to the `check_stata()` command.)

yuchuan2016 commented:
@lboxell , thanks! It looks much better now! I'll go ahead first merge master into this branch.

yuchuan2016 commented:
Note: `requests` module has not been added to the `gslab_python` dependencies. See [this comment](https://github.com/gslab-econ/template/issues/19#issuecomment-260149058).

arosenbe commented:
@yuchuan2016. A later commit must have overwritten my change to install the requests module. See [here](https://github.com/gslab-econ/gslab_python/blob/ff832fb3132f1535f08e1a093c6f7ef65b0cae7c/setup.py) for my implementation in `setup.py`.

yuchuan2016 commented:
Thanks a lot for your message @arosenbe ! Would you mind updating the current `gslab_python` `setup.py` to incorporate this again? Thanks!

arosenbe commented:
Sure thing @yuchuan2016! And thanks for bringing it to my attention. It's done in gslab-econ/gslab_python#43 and PR  gslab-econ/gslab_python#48.

yuchuan2016 commented:
Per discussion, the location of `Metropolis beamer theme` is not fixed on Windows. So instead of checking it explicitly, the `setup.py` will throw a warning that this particular requirement has not been checked when running on Windows. When it is not installed, `talk/slide.lyx` will not compile successfully.

@lboxell , another thing is that, in `SConstruct`, before running `setup_test`, we import `gslab_scons`. But technically we need to check whether `gslab_tools` have been installed or not at the beginning. Do you think this is an issue? Besides that, could you take a look at the current version to see if everything works well? Thanks!

lboxell commented:
@yuchuan2016,

Good point about the import order in the SConstruct. What do you think about the change I just made to address that?

If you agree, I think we can merge into master and then open up the task to address any redundancies in gslab_scons and setup.py

yuchuan2016 commented:
@lboxell , 

Great, thanks! I'll merge it to the master.

yuchuan2016 commented:
Hi @lboxell ,

Since @stanfordquan is working on refactoring`build_stata` related issues [here](https://github.com/gslab-econ/gslab_python/commit/99b4fb822da337bba8f5883df47285c39eb47156),  so it may be better to remove any redundancies after the refactor.

lboxell commented:
@yuchuan2016,

Sounds good. You can just make a task in gslab_python for this and note that it might want to wait until build_stata is refactored. Or flag @stanfordquan to implement this as a part of his task. Then I think you can close this task.

stanfordquan commented:
Noted. @yuchuan2016 you can close the task! I'll remove the redundancies as part of gslab_python/#28 gslab_python#47

yuchuan2016 commented:
Thanks a lot @stanfordquan!!
